File: packages/core/src/routes/audit.ts
/**
 * Audit routes — chain tracing and integrity checking.
 */

import { Hono } from 'hono'
import type { Services } from '../services'
import { mapStatusCode } from './server'

export function auditRoutes(services: Services) {
  const app = new Hono()

  app.get('/audit/chain/:perspectiveId', async (c) => {
    const result = await services.audit.traceChain(c.req.param('perspectiveId'))
    if (!result.ok) return c.json({ error: result.error }, mapStatusCode(result.error.code))
    return c.json({ data: result.data })
  })

  app.get('/audit/integrity/:perspectiveId', async (c) => {
    const result = await services.audit.checkIntegrity(c.req.param('perspectiveId'))
    if (!result.ok) return c.json({ error: result.error }, mapStatusCode(result.error.code))
    return c.json({ data: result.data })
  })

  return app
}

File: crates/forge-server/src/routes/audit.rs
//! Audit routes — chain tracing and integrity checking.
//!
//! Mirrors `packages/core/src/routes/audit.ts`.

use axum::extract::{Path, State};
use axum::routing::get;
use axum::{Json, Router};

use forge_core::{ChainTrace, IntegrityReport};
use forge_sdk::services::AuditService;

use crate::db::with_conn;
use crate::error::ApiError;
use crate::response::ApiData;
use crate::state::SharedState;

// ── Handlers ────────────────────────────────────────────────────────

async fn trace_chain(
    State(state): State<SharedState>,
    Path(perspective_id): Path<String>,
) -> Result<Json<ApiData<ChainTrace>>, ApiError> {
    let result = with_conn(&state, move |conn| {
        AuditService::trace_chain(conn, &perspective_id)
    })
    .await?;
    Ok(Json(ApiData { data: result }))
}

async fn check_integrity(
    State(state): State<SharedState>,
    Path(perspective_id): Path<String>,
) -> Result<Json<ApiData<IntegrityReport>>, ApiError> {
    let result = with_conn(&state, move |conn| {
        AuditService::check_integrity(conn, &perspective_id)
    })
    .await?;
    Ok(Json(ApiData { data: result }))
}

// ── Router ──────────────────────────────────────────────────────────

pub fn router() -> Router<SharedState> {
    Router::new()
        .route("/audit/chain/{perspectiveId}", get(trace_chain))
        .route("/audit/integrity/{perspectiveId}", get(check_integrity))
}
